Feature/pin code gen

KarmaKamikaze opened the pull request.

KarmaKamikaze committed:
PinDclNode has been added. In addition, CodeGen uses a special visitor to grab the pin int in a temp variable.

KarmaKamikaze committed:
Output code gen to optional string, which must be included in Setup.

KarmaKamikaze committed:
Added saved pinMode from variables scope to the setup code emits.

KarmaKamikaze committed:
Fixed string program, where only one pin could be declared.

mirakst commented on STEP/AST/AstPrintVisitor.cs:
```diff
@@ -97,6 +97,11 @@ public void Visit(VarDclNode node)
 
+    public void Visit(PinDclNode n)
+    {
+        throw new NotImplementedException();
```
Implementer printer!

KarmaKamikaze replied:
👍

mirakst commented on STEP/CodeGeneration/CodeGenerationVisitor.cs:
```diff
@@ -9,7 +9,8 @@ public class CodeGenerationVisitor : IVisitor
     private readonly StringBuilder _stringBuilder = new();
     private string Output => _stringBuilder.ToString();
-    
+    private StringBuilder _optionalPinMode = new();
```
```suggestion
    private StringBuilder _pinSetup = new();
```

KarmaKamikaze committed:
Rename _optionalPinMode to _pinSetup

mirakst commented on STEP/TypeVisitor.cs:
```diff
@@ -221,6 +221,11 @@ public void Visit(ArrLiteralNode n)
+    {
+        throw new NotImplementedException();
+    }
+
```
jeg hopper lige på branchen og fikser det her

mirakst replied:
Det er fixed!

Sarmisuper reviewed (approved):
LGTM
LIttel Guy Tisse Mand

Commit:
Added type checking for PinDclNodes

Tests still succeed

mirakst committed:
Merge branch 'dev' into feature/PinCodeGen

Commit:
Fixed small oversight from merge

mirakst reviewed (approved):
Jeg har ordnet type visitor delen samt fixed tests - det hele kører!
Jeg har også fixed den merge conflict der kom af det.
Hvis du får implementeret print metoden, så er det fantastisk!

KarmaKamikaze committed:
PinDclNode printer implemented

KarmaKamikaze committed:
Merge branch 'feature/PinCodeGen' of https://github.com/KarmaKamikaze/STEP into feature/PinCodeGen

Commit:
Removed symbol table entry from varsnode visit method

Commit:
Fixed builder method for PinDclNode

Commit:
Correctly sets type of PinDclNode

KarmaKamikaze committed:
Print PinDclNode's Id node's type instead of own type in pretty printer.

damniko commented on STEP/AST/Nodes/PinDclNode.cs:
```diff
@@ -0,0 +1,17 @@
+﻿namespace STEP.AST.Nodes;
+
+public class PinDclNode : VarDclNode {
+    public PinType PinType { get; set; } = new();
```
Skal den her ikke bare være `Type` som den nedarver fra `VarDclNode`?
Eftersom `PinType` nedarver fra `Type` bør det være muligt

KarmaKamikaze replied:
Det burde være fikset i 38df0ce3ed3bf6bc079f9377c61f82cc7623ed25

damniko commented on STEP/CodeGeneration/CodeGenerationVisitor.cs:
```diff
@@ -571,4 +576,26 @@ public void Visit(ElseIfNode n)
         }
     }
+
+    public void Visit(PinDclNode n)
+    {
+        PinCodeVisitor pinVisitor = new PinCodeVisitor();
+        /* The emitted code will be stored in a temporary variable. If any pins are declared, we know
+         * it must be declared in the variables scope, which is always visited first. Once the code generation
+         * reaches the Setup scope, the temporary variables will be used to insert this code in the correct place.
+         */
+        _pinSetup.Append("pinMode(");
+        n.Left.Accept(pinVisitor);
+        _pinSetup.Append(pinVisitor.GetPinCode());
+        switch (n.PinType.Mode)
+        {
+            case PinMode.INPUT:
+                _pinSetup.Append("INPUT");
+                break;
+            case PinMode.OUTPUT:
+                _pinSetup.Append("OUTPUT");
+                break;
+        }
+        _pinSetup.Append(");\n");
+    }
```
![image](https://user-images.githubusercontent.com/71791903/165292430-dd5822a9-db91-4deb-b386-d7c5fcab462b.png)

damniko commented on STEP/TypeVisitor.cs:
```diff
@@ -194,26 +194,32 @@ public class TypeVisitor : IVisitor {
-                    _pinTable.RegisterPin(n.Left.Type.ActualType, pinVal);
-                    break;
-            }
+        if (n.Left.Type == n.Right.Type) {
```
Bør den eksplicit tjekke om de er andet end analogpin og digitalpin?

KarmaKamikaze replied:
Det burde være nok at tjekke de to objects for equality, right? Hvis de er andet end analogpin eller digitalpin, så bør de ikke være equal.

damniko reviewed (approved):
very beautiful craftmanship

Commit:
Allowed empty array literals

Code generation will not emit an initializer, arduino default behavior is to initialize to default values.

KarmaKamikaze committed:
Changed PinType on PinDclNode to Type.

mirakst committed:
Merge pull request #20 from KarmaKamikaze/feature/EmptyArrLit

Feature/empty arr lit
